wip: fix(virtual-scroll): make horizontal scrolling work in RTL

mmalerba opened the pull request.

crisbeto commented on src/cdk/scrolling/scrolling-module.ts:
```diff
@@ -14,8 +15,9 @@ import {CdkVirtualForOf} from './virtual-for-of';
 
 @NgModule({
-  imports: [PlatformModule],
+  imports: [BidiModule, PlatformModule],
```
I don't think you need to import the `BidiModule` since the `Directionality` is provided at the root.

mmalerba replied:
I think it definitely needs to be in `exports`, so that we detect native `dir="rtl"` attrs as directives. I'm not sure if it needs to be in `imports`, I see a lot of other modules doing it though. @jelbourn do you know if having it in `imports` is still necessary?

crisbeto commented on src/cdk/scrolling/virtual-scroll-viewport.spec.ts:
```diff
@@ -629,3 +706,53 @@ class FixedSizeVirtualScroll {
+  @Input() itemSize = 50;
+  @Input() bufferSize = 0;
+  @Input() items = Array(10).fill(0).map((_, i) => i);
+  @Input() trackBy;
```
This one's missing a type.

crisbeto commented on src/cdk/scrolling/virtual-scroll-viewport.ts:
```diff
@@ -270,13 +278,20 @@ export class CdkVirtualScrollViewport implements OnInit, OnDestroy {
 
+    // For a horizontal viewport in a right-to-left language we need to calculate what `scrollRight`
+    // would be.
+    offset = this.orientation == 'horizontal' && this._dir && this._dir.value == 'rtl' ?
```
It looks like you're repeating `Math.max(0, this._totalContentSize - this._viewportSize - offset)` at least in a couple of places. Consider moving it into a helper.

crisbeto commented on src/demo-app/demo-app/demo-app.html:
```diff
@@ -41,7 +41,7 @@ <h1>Angular Material Demos</h1>
     </mat-toolbar>
 
-    <div #root="dir" dir="ltr" class="demo-content">
+    <div #root="dir" dir="rtl" class="demo-content">
```
Should this be reverted?

jelbourn commented on src/cdk/scrolling/virtual-scroll-viewport.ts:
```diff
@@ -243,8 +247,12 @@ export class CdkVirtualScrollViewport implements OnInit, OnDestroy {
+    // For a horizontal viewport in a right-to-left language we need to translate along the x-axis
+    // in the negative direction.
+    const axisDirection = this.orientation == 'horizontal' && this._dir && this._dir.value == 'rtl'
+        ? -1 : 1;
```
Since you check them a couple of time, I would add `_isHorizontal` and `_isRtl` methods. Then this would just look like
```ts
// For a horizontal viewport in a right-to-left language we need to translate along the x-axis
// in the negative direction.
const axisCoefficient = this._isHorizontal && this._isRtl() ? -1 : 1;
```

mmalerba commented:
@jelbourn @crisbeto PTAL, I refactored this to have the `CdkVirtualScrollViewport` extend `CdkScrollable` and take advantage of the RTL logic I already added to `CdkScrollable`

jelbourn commented on src/cdk/scrolling/virtual-scroll-viewport.ts:
```diff
@@ -238,8 +244,12 @@ export class CdkVirtualScrollViewport implements OnInit, OnDestroy {
+    // For a horizontal viewport in a right-to-left language we need to translate along the x-axis
+    // in the negative direction.
+    const axisDirection = this.orientation == 'horizontal' && this.dir && this.dir.value == 'rtl'
+        ? -1 : 1;
```
How about 
```ts
const isHorizontal = this.orientation == 'horizontal';
const isRtl = this.dir && this.dir.value == 'rtl';
const axisDirection = isHorizontal && isRtl ? -1 : 1;
```

mmalerba commented on src/cdk/scrolling/scrollable.ts:
```diff
@@ -48,21 +48,21 @@ export class CdkScrollable implements OnInit, OnDestroy {
-              @Optional() private _dir?: Directionality) {}
+  constructor(protected elementRef: ElementRef<HTMLElement>,
+              protected scrollDispatcher: ScrollDispatcher,
+              protected ngZone: NgZone,
```
I think we want people to extend their own components from this if they act as a viewport. Therefore the `protected` properties are kind of a part of the public API and should not be underscored
